thunder/benchmarks/benchmark_peft.py

```diff
@@ -0,0 +1,724 @@
+                        config,
+                        torch_dtype=torch.bfloat16,
+                        trust_remote_code=trust_remote_code,
+                    )
+                except Exception as e:
+                    logger.warning(
+                        f"Could not create model as AutoModel, trying LlavaForConditionalGeneration: {str(e)}"
+                    )
+                    from transformers import LlavaForConditionalGeneration
+
+                    model = LlavaForConditionalGeneration._from_config(
+                        config,
+                        torch_dtype=torch.bfloat16,
+                    )
+
+        if model is None:
+            raise RuntimeError("Failed to create model on meta device")
+
+        if verbose:
+            logger.info("Model created on meta device")
+
+        return model
+
+    except Exception as e:
+        logger.error(f"Error in model configuration: {str(e)}")
+        logger.error(f"GPU memory at error: {torch.cuda.memory_allocated()/1024**3:.2f}GB")
+        raise
+
+
+def setup_lora(model: torch.nn.Module, verbose: bool = False) -> torch.nn.Module:
+    """Apply LoRA to the model."""
+    if verbose:
+        logger.info("Applying LoRA to model")
+
+    lora_config = LoraConfig(
+        r=16,  # rank
+        target_modules="all-linear",  # See: https://huggingface.co/docs/peft/package_reference/lora#peft.LoraConfig.target_modules
+        lora_alpha=32,
+        task_type="CAUSAL_LM",
+    )
+
+    model = get_peft_model(model, lora_config)
+
+    # Freeze all parameters except LoRA
+    for name, param in model.named_parameters():
+        if "lora" not in name.lower():
+            param.requires_grad = False
+        else:
+            param.requires_grad = True
+            if verbose:
+                logger.info(f"Setting gradient for LoRA parameter: {name}")
+
+    if verbose:
+        logger.info("LoRA applied to model")
+
+    return model
+
+
+def setup_fsdp2(model: torch.nn.Module, devices: int, verbose: bool = False) -> torch.nn.Module:
+    """Apply FSDP2 to the model with ZeRO-3 style sharding."""
+    if devices <= 1:
+        return model
+
+    if verbose:
+        logger.info("Applying FSDP2 to model with ZeRO-3 style sharding")
+
+    mesh = DeviceMesh("cuda", torch.arange(devices), mesh_dim_names=("dp",))
+
+    # Configure model for static shapes before FSDP2
+    if hasattr(model, "config"):
+        model.config.use_cache = True
+        model.config.max_position_embeddings = args.seq_length
+        if verbose:
+            logger.info(f"Configured model for static shapes with sequence length: {args.seq_length}")
+
+    # Apply FSDP2 with ZeRO-3 style sharding
+    # First wrap individual layers
+    for name, module in model.named_modules():
+        if isinstance(module, (torch.nn.Linear, torch.nn.Embedding)):
+            if verbose:
+                logger.info(f"Wrapping layer {name} with FSDP2")
+            fully_shard(module, mesh=mesh)
```

Comment:
just for the consistency with the below
```suggestion
            fully_shard(module, mesh=mesh, reshard_after_forward=True)
```